### PR TITLE
feat(state): keep global state of pinned buffers for session persistence

### LINKED ISSUE #15: Persist pinned buffers
Is it possible to persist the list of pinned buffers using a plugin like `auto-session`?

### PR DESCRIPTION
**Pull request by Nimjii**
Hey again!

This should fix #15.

I also made `set_all` a local function, as it seemed kind of useless to export it (I can remove that change tho, if you don't like it). 
I won't lie, I don't use any session plugins, so I am not exactly sure on how they operate. If this has any issues, feel free to point them out.

### CODE REVIEW

**Comment by al-ce** [comment]
Thanks for making this PR! This would be a nice feature and I think your changes are elegant and easy to follow.

I should note that I'm testing this by using `mksession` as my session tool (as opposed to a plugin that uses JSON or something else). I am executing `mksession` before I quit and I am using this command to restore the session: `nvim -u minimal.lua -S Session.vim` This is my minimal config.

<details>

<summary>minimal config</summary>

```lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable", -- latest stable release
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)

vim.go.sessionoptions = "buffers,curdir,tabpages,winsize,help,globals"

require("lazy").setup({
  {"Nimjii/hbac.nvim",
    opts = {},
  },
}, {})
```
</details>
 
The first thing that caught my eye is that `vim.g.hbac_buffers` is all lowercase, and if you look at point 2 in `:h mksession` it notes
 
> 2. Restores global variables that start with an uppercase letter and contain
   at least one lowercase letter, if 'sessionoptions' contains "globals".

The pins weren't restoring for me when I reloaded my session with the current commit. After changing the global variable to `Hbac_buffers` throughout the project, the pins were being restored (with some caveats, see below). Does it work for you as it currently is? Do you have a session method you're using?

Second, the `unpin_all` function wasn't working for me anymore, I think it just needs an early return when `pin_value` is false. Maybe we just want to separate the logic instead of calling set_all for both pin_all and unpin_all at this point?

With these two changes, it mostly works! But if I have more buffers open than my threshold setting, even if they're all pinned, and if I save the session, quit, and try to reload it, the pinned bufs aren't as i'd expect. There's kind of a mix of conditions that I'm trying out (threshold value, amount of buffers opened/pinned/unpinned on session save and quit, whether my config sets autoclose to false, whether I re-pin the buffers on the session restore and then quit and reload the session, etc.), but I'm afraid of adding more confusion to the mix with my interpretations or possible fixes since I don't fully understand it, so I'm going to sit on it for a few days. Also maybe it's just me. Can you please experiment with those when you have a chance? (I suggest setting a low threshold for testing). Let me know if my comments are helpful or if I'm missing something and it should be working.

Thanks for working on this, I'm excited about this feature! We should also add a note in the README that sessionoptions should include `globals` for this to work

File: lua/hbac/command/actions.lua [hunk 1]
```diff
@@ -19,20 +19,27 @@ M.toggle_pin = function(bufnr)
 	return bufnr, pinned_state
 end
 
-M.set_all = function(pin_value)
+local set_all = function(pin_value)
 	local utils = require("hbac.utils")
 	local buflist = utils.get_listed_buffers()
+	local pins = utils.int_explode(vim.g.hbac_buffers or "")
+	if not pin_value then
+		vim.g.hbac_buffers = ""
```
Comment (al-ce) [inline lua/hbac/command/actions.lua:27]: I think this line should have the early return (or we just make pin_all and unpin_all do their own logic?). I think line 34 is overwriting this with the buflist equivalent even on `unpin_all`

Comment (Nimjii) [reply lua/hbac/command/actions.lua:27]: Yeah I there should have definitely been a return there (there is now). Tbh I forgot to test `unpin_all` (I did test `pin_all` ^^).

**Comment by Nimjii** [comment]
Thanks for all the feedback. I will get to the more thorough testing, when I find the time.

**Comment by axkirillov** [comment]
@Nimjii hi, what's the status here? are you still planning to merge?

**Comment by axkirillov** [comment]
@Nimjii seems like you abandoned the idea, so I'll close it for now

**Pull request closed without merging**